File: workspace/server/src/domain/tasks/xp.ts
import type { Difficulty, Size } from "../../db/task-types.ts";

const BASE_XP: Record<Difficulty, number> = {
  low: 10,
  normal: 20,
  high: 40,
};

const SIZE_MULT: Record<Size, number> = {
  tiny: 0.5,
  small: 0.75,
  medium: 1.0,
  large: 1.5,
};

export function computeTaskXp(difficulty: Difficulty, size: Size): number {
  return Math.round(BASE_XP[difficulty] * SIZE_MULT[size]);
}

export function xpForLevel(level: number): number {
  return (level - 1) ** 2 * 100;
}

File: backend/internal/task/xp.go
package task

// XPPerLevel is the canonical XP-per-level constant for P05+ progression math.
// ponytail: profile/service.go (P04, frozen) keeps its own unexported xpPerLevel=1000;
// unify into a shared package if a third consumer appears.
const XPPerLevel = 1000

// Base XP by task complexity.
var complexityXP = map[string]int{
	"tiny":   50,
	"small":  100,
	"medium": 200,
	"large":  350,
}

const highPriorityBonus = 50

// BaseXP computes the non-negative base XP a task would grant on completion.
// Unknown complexity yields 0 (no XP). Negative results are impossible.
// P06 owns actual granting + multipliers; P05 only exposes the helper.
func BaseXP(complexity string, priority string) int {
	xp := complexityXP[complexity]
	if priority == "high" {
		xp += highPriorityBonus
	}
	return xp
}
